code/Serving_API.py

```diff
@@ -124,9 +146,62 @@ def calculate_shortest_path_distance(G, shortest_path_cities):
 
         total_distance += G[city1][city2]['weight']
         # Only append the coordinates of the source city here
-        coordinate_list.append(G.nodes[city1]['coord'])
-        wkt_list.append(G[city1][city2]['path_wkt'])
-        cable_type_list.append(G[city1][city2]['cable_type'])
+        # for each edge, find all the available asn nodes that are close to the edge
+        insertable_asn = []
+        for lat, lon in asn_nodes:
+            if lat != 'NULL' and lon != 'NULL':
+                result = find_closest_paths(
+                    lat, lon, G[city1][city2]['path_wkt'], 5)
+                if not result.empty:
+                    insertable_asn.append((lat, lon))
+        
+        # has asn nodes that are close to the edge
+        if len(insertable_asn) > 0:
+            
+            item_distance_pairs = []
+            
+            # sort the asn nodes by distance to the start point of the edge for following cut option
+            for item in insertable_asn:
+                distance = G[city1][city2]['path_wkt'].project(
+                    Point(item[1], item[0]))
+                item_distance_pairs.append((item, distance))
+
+            sorted_item_distance_pairs = sorted(
+                item_distance_pairs, key=lambda x: x[1])
+
+            linestring_to_be_cut = G[city1][city2]['path_wkt']
+            linestring_to_be_cut_type = G[city1][city2]['cable_type']
+            
+            # append the start point of the edge
+            coordinate_list.append(G.nodes[city1]['coord'])
+            # append the cable type of the edge for the first segment
+            cable_type_list.append(linestring_to_be_cut_type)
+            
+            for item, distance in sorted_item_distance_pairs:
+                lat, lon = item
+                
+                splitted = cut(linestring_to_be_cut, linestring_to_be_cut.project(
+                    Point(lon, lat)), Point(lon, lat))
```

Comment:
I don't think this is ideal. Point(lon, lat) may not necessarily be on the linestring, so you're cutting correctly, but not adding the new point correctly. I think you want to save the distance and use [`interpolate`](https://shapely.readthedocs.io/en/stable/reference/shapely.LineString.html#shapely.LineString.interpolate) to find the tangent point.

Reply:
This is fixed in the new cut function.